### PR TITLE
resolved_ts: track pending lock memory usage

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: ref #14864

What's Changed:

```commit-message
* Fix resolved ts OOM caused by adding large txns locks to `ResolverStatus`.
* Add initial scan backoff duration metrics.
```

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/resolved_ts/src/endpoint.rs [hunk 1]
```diff
@@ -53,10 +53,138 @@ enum ResolverStatus {
...
+            return;
+        }
+
+        // Free memory quota used by pending locks and unlocks.
+        let mut bytes = 0;
+        let num_locks = locks.len();
+        for lock in locks {
+            bytes += lock.heap_size();
+        }
+        if bytes > ON_DROP_WARN_HEAP_SIZE {
```
Comment (tonyxuqqi) [inline components/resolved_ts/src/endpoint.rs:80]: Should we take this chance to log the current memory_quota value?

File: components/resolved_ts/src/endpoint.rs [hunk 2]
```diff
@@ -101,129 +240,72 @@ impl ObserveRegion {
...
-                            });
-                            assert!(
-                                *tracked_index < *index,
-                                "region {}, tracked_index: {}, incoming index: {}",
-                                self.meta.id,
-                                *tracked_index,
-                                *index
+        if matches!(self.resolver_status, ResolverStatus::Pending { .. }) {
+            self.resolver_status
+                .track_change_log(self.meta.id, change_logs)
```
Comment (tonyxuqqi) [inline components/resolved_ts/src/endpoint.rs:245]: Question: Why is this change needed?
It separates the implementation of ResolverStatus::Pending from other state, which may make the readability worse.

File: components/resolved_ts/src/resolver.rs [hunk 3]
```diff
@@ -68,10 +68,12 @@ impl Drop for Resolver {
             bytes += self.lock_heap_size(key);
         }
         if bytes > ON_DROP_WARN_HEAP_SIZE {
-            warn!("drop huge resolver";
+            warn!("drop huge Resolver";
```
Comment (Connor1996) [inline components/resolved_ts/src/resolver.rs:71]:
```suggestion
            warn!("drop huge resolver";
```

**Review by Connor1996 — APPROVED** [review]
LGTM

**Review by tonyxuqqi — APPROVED** [review]
